### ISSUE TITLE
Can't provide a PPM and PWM output at the same time

### ISSUE THREAD

**Issue by DeLavel** [madhephaestus/ESP32Servo#74]
Hi, hope that someone might help - should be simple but I cannot figure out how to provide a pwm and ppm (servo) signal at the same time.
This is for an ESP32C3, and I need a servo output on one pin, with a pwm signal on a second pin.  It works, but the pwm frequency takes that of the servo - 50Hz.  I need it to be 10KHz.  I understand that it's a timer clash - but how to fix it?  Thanks for looking..

 /* modified from the examples in the  ESP32 Servo Library
  it works - but the pwm is only 50hz...
  */
 
 #include <ESP32Servo.h>

 // create  servo object 
 Servo servo1;

 // Published values for SG90 servos; adjust if needed
 int minUs = 1000;
 int maxUs = 2000;
 
 
 int servo1Pin = 7;
 int LedPin=21;

int channel=0; // timer channel return value
 int pos = 0;      // position in degrees
 ESP32PWM pwm;

 void setup() {
     // Allow allocation of all timers
      ESP32PWM::allocateTimer(0);
      ESP32PWM::allocateTimer(1);
      ESP32PWM::allocateTimer(2);
      ESP32PWM::allocateTimer(3);
     Serial.begin(9600);
     servo1.setPeriodHertz(50);      // Standard 50hz servo

     pwm.attachPin(LedPin, 10000);//10khz  // important to do this before the servo attach, or the servo goes nuts
 Serial.println("pwm attached");

 channel=   servo1.attach(servo1Pin, minUs, maxUs);
 Serial.print("servo1 on timer channel: "); Serial.println(channel);
 
 }
 
 void loop() {
 //pwm.writeScaled(0.5);
     for (pos = 0; pos <= 180; pos += 1) { // sweep from 0 degrees to 180 degrees
         // in steps of 1 degree
         servo1.write(pos);
         
         delay(1);             // waits 20ms for the servo to reach the position
     }
     for (pos = 180; pos >= 0; pos -= 1) { // sweep from 180 degrees to 0 degrees
         servo1.write(pos);
         delay(1);
     }
 
  
// fade the LED on thisPin from off to brightest:
	for (float brightness = 0; brightness <= 0.5; brightness += 0.001) {
		// Write a unit vector value from 0.0 to 1.0
		pwm.writeScaled(brightness);
		delay(2);
        
	}
	
	for (float brightness = 0.5; brightness >= 0; brightness -= 0.001) {
        	pwm.writeScaled(brightness);
		delay(2);
	}

 
     delay(1000);
 
 }
 
`

**Comment by madhephaestus** [comment]
Attach the servo first, that will set up the timer to the 50hz for servo.  Then run the  pwm attach which call allocateNext and pick up the next timer. I think the Servo class will force the allocated PWM timer to 50 when attached, but allocate next should pick up the next one since the fequency is the parameter

**Comment by DeLavel** [comment]
Thanks for the suggestion, but I think that this is one of the things that I did try.  That said, I have tried so many combinations and workarounds that I lost track - am at work atm, but will give this a go later - thanks again.

**Comment by DeLavel** [comment]
So, I tried adding 4 servos before attaching the pwm  - and whilst I get a 10KHz signal for pwm - the servo goes nuts, showing a frequency bouncing around 30KHz
I've tried this, earlier, having found a comment about adding dummy servos.  The sketch returns the following when run - perhaps I misunderstand, but the servo attach method - a return value of 0 is 'failure to attach'?  Thanks again for any advice...
18:42:14.598 -> ESP-ROM:esp32c3-api1-20210207
18:42:14.598 -> servo1 on timer channel: 0
18:42:14.598 -> servo2 on timer channel: 0
18:42:14.598 -> servo3 on timer channel: 0
18:42:14.598 -> servo4 on timer channel: 0
18:42:14.598 -> pwm attached

**Comment by madhephaestus** [comment]
Can you put the full code on git so i can see the problem fully in context?

Also try getting rid of the frequency setting of the servo object, for standard frequency that is redundant, and may be related to the cross talk issue.

Creating 4 dummy servos and attaching them on pins, then attach the pwm, then detach the dummy 3 servos.

**Comment by DeLavel** [comment]
Thanks for looking at this.  I tried your suggestions but they result in mad frequencies found at both ppm and pwm outputs..

`

 /* modified from the examples in the 
  * ESP32 Servo Library
  it works - but the pwm is only 50hz...
  saved out as 'experimental'
  */
 
 #include <ESP32Servo.h>


 // create  servo objects
 Servo servo1;
 Servo servo2;
 Servo servo3;
 Servo servo4;

 // Published values for SG90 servos; adjust if needed
 int minUs = 1000;
 int maxUs = 2000;
 
 // These are all GPIO pins on the ESP32
 // Recommended pins include 2,4,12-19,21-23,25-27,32-33 
 int servo1Pin = 7;  // real one
 int servo2pin=12;
 int servo3pin=13;
 int servo4pin=14;

 int LedPin=21;
 //int dummyLedPin=4;

int channel=0; // timer channel return value
 int pos = 0;      // position in degrees
 ESP32PWM pwm;
 //ESP32PWM dummypwm;

 void setup() {
  
     // Allow allocation of all timers
      ESP32PWM::allocateTimer(0);
      ESP32PWM::allocateTimer(1);
      ESP32PWM::allocateTimer(2);
      ESP32PWM::allocateTimer(3);

     Serial.begin(9600);



// servo1.setPeriodHertz(50);      // Standard 50hz servo
// servo2.setPeriodHertz(50); 
// servo3.setPeriodHertz(50); 
// servo4.setPeriodHertz(50); 

 channel=   servo1.attach(servo1Pin, minUs, maxUs);
 Serial.print("servo1 on timer channel: "); Serial.println(channel);
 
 channel=   servo2.attach(servo2pin, minUs, maxUs);
 Serial.print("servo2 on timer channel: "); Serial.println(channel);

 channel=   servo3.attach(servo3pin, minUs, maxUs);
 Serial.print("servo3 on timer channel: "); Serial.println(channel);

 channel=   servo4.attach(servo4pin, minUs, maxUs);
 Serial.print("servo4 on timer channel: "); Serial.println(channel);


pwm.attachPin(LedPin, 10000);//10khz  // important to do this before the servo attach!!!!
 Serial.println("pwm attached");

servo2.detach();
servo3.detach();
servo4.detach();


 }
 

 void loop() {
 pwm.writeScaled(0.5);  // led on - not bothered about fading, 
 //I just want to measure the pwm frequency with the scope
   
 sweep(); // check that the servo is working ok
 
     delay(1000);
 
 }
 

// * ledc: 0 => Group: 0, Channel: 0, Timer: 0
// * ledc: 1 => Group: 0, Channel: 1, Timer: 0
// * ledc: 2 => Group: 0, Channel: 2, Timer: 1
// * ledc: 3 => Group: 0, Channel: 3, Timer: 1
// * ledc: 4 => Group: 0, Channel: 4, Timer: 2
// * ledc: 5 => Group: 0, Channel: 5, Timer: 2
// * ledc: 6 => Group: 0, Channel: 6, Timer: 3
// * ledc: 7 => Group: 0, Channel: 7, Timer: 3


void sweep(){
  for (pos = 0; pos <= 180; pos += 1) { // sweep from 0 degrees to 180 degrees
         // in steps of 1 degree
         servo1.write(pos);
         
         delay(1);             // waits 20ms for the servo to reach the position
     }
     for (pos = 180; pos >= 0; pos -= 1) { // sweep from 180 degrees to 0 degrees
         servo1.write(pos);
         delay(1);
     }
 
  

}

void fade() {
// fade the LED on thisPin from off to brightest:
	for (float brightness = 0; brightness <= 0.5; brightness += 0.001) {
		// Write a unit vector value from 0.0 to 1.0
		pwm.writeScaled(brightness);
		delay(2);
        
	}
	
	for (float brightness = 0.5; brightness >= 0; brightness -= 0.001) {
        	pwm.writeScaled(brightness);
		delay(2);
	}
}

`

**Comment by zhengyangliu** [comment]
https://github.com/espressif/arduino-esp32/blob/31d22e6ed05a26a7fd7765b9ae42fd292559690f/cores/esp32/esp32-hal-ledc.c#L185 
It is certain that the ledcAttach of core3.x does not automatically allocate timers according to frequency and resolution, it simply detects whether the channel is used. It seems that ledcAttachChannl needs to be used to fix this bug.
